### DISCUSSION TITLE
Unsubscribe handling

### DISCUSSION THREAD

**Discussion by ShadowWolf308** [enisdenjo/graphql-ws#593] [Q&A] [answered]
How do i prevent the `onNext` from being called for a subscription after receiving a `complete` messsage from the client?

I have tried removing the iterator from `context.subscriptions` but it still calls the `onNext` regardless.
The `onComplete` listener in `useServer` also does not get called after receiving a `complete` message from the client, i do receive this message on `webSocket.onMessage` however.

**Comment by enisdenjo** [comment]
The `onNext` callback is [bound to the subscription async iterable](https://github.com/enisdenjo/graphql-ws/blob/2c603adc241395d616e41ed45e5d60ef551600d2/src/server.ts#L819-L821). You can prevent it from being called by properly disposing/returning your subscription iterable which will in turn break that loop and stop invoking the `onNext` callback.

Also, the fact that the `onComplete` callback is not invoked prooves the point that your iterable is not being completed properly.

**Reply by ShadowWolf308** [reply]
And how do i properly dispose of that iterable?

i tried using `delete context.subscriptions` and that didn't work.

i have been looking for a while to see if i could find any documentation on disposing of subscriptions properly but couldn't find anything.

**Reply by enisdenjo** [reply, marked as answer]
The issue is not in graphql-ws at all, there's nothing you should/can do there. The issue is with your subscriptions in the graphql resolvers.

**Reply by ShadowWolf308** [reply]
I found why my unsubscribes aren't working, thanks for the support. it was indeed in the subscription of the resolver.

**Reply by onionhammer** [reply]
> I found why my unsubscribes aren't working, thanks for the support. it was indeed in the subscription of the resolver.


Why?

**Reply by ShadowWolf308** [reply]
> Why?

The asyncIterator has a return function to handle closing the iterator, that function was somehow not always set on my end and could therefore never really close the iterator, that ended up preventing my unsubscribes from not working, making sure that that function is set and resolves fixed it for me
